File: android-native/app/src/main/java/com/example/kpkn/screens/workout/WorkoutCoachMessages.kt
package com.example.kpkn.screens.workout

data class CoachMessage(
    val key: String,
    val title: String,
    val body: String,
    val severity: CoachSeverity,
    val action: CoachAction? = null,
)

enum class CoachSeverity { INFO, WARNING, DANGER, SUCCESS }

enum class CoachAction { REDUCE_INTENSITY, SKIP_EXERCISE, EXTEND_REST, STAY_THE_COURSE }

object WorkoutCoachMessages {
    enum class DrainLevel { XS, LOW, AVERAGE, HIGH }
    enum class ReadinessLevel { XS, LOW, AVERAGE, HIGH }

    private data class CoachKey(
        val drain: DrainLevel,
        val readiness: ReadinessLevel,
        val sessionProgress: SessionPhase,
    )

    private enum class SessionPhase { EARLY, MID, LATE }

    private val COACH_MESSAGES: Map<CoachKey, CoachMessage> = buildMap {
        val drainLevels = DrainLevel.entries
        val readinessLevels = ReadinessLevel.entries
        val phases = SessionPhase.entries

        for (drain in drainLevels) {
            for (readiness in readinessLevels) {
                for (phase in phases) {
                    val (severity, title, body, action) = resolve(drain, readiness, phase)
                    put(CoachKey(drain, readiness, phase), CoachMessage(
                        key = "${drain.name}_${readiness.name}_${phase.name}",
                        title = title,
                        body = body,
                        severity = severity,
                        action = action,
                    ))
                }
            }
        }
    }

    private fun resolve(
        drain: DrainLevel,
        readiness: ReadinessLevel,
        phase: SessionPhase,
    ): CoachMessageComponents {
        val isHighDrain = drain == DrainLevel.HIGH || drain == DrainLevel.AVERAGE
        val isLowReadiness = readiness == ReadinessLevel.XS || readiness == ReadinessLevel.LOW
        val isLateSession = phase == SessionPhase.LATE

        return when {
            isHighDrain && isLowReadiness && isLateSession -> CoachMessageComponents(
                severity = CoachSeverity.DANGER,
                title = "Cuidado: fatiga acumulada + baja disponibilidad",
                body = "Has acumulado fatiga significativa y tu readiness está baja. Considera terminar ejercicios principales aquí y guardar energía.",
                action = CoachAction.REDUCE_INTENSITY,
            )
            isHighDrain && isLowReadiness && !isLateSession -> CoachMessageComponents(
                severity = CoachSeverity.WARNING,
                title = "Fatiga en aumento, readiness limitada",
                body = "El sistema nervioso está trabajando fuerte y la recuperación de hoy no es óptima. Prioriza las series más importantes.",
                action = CoachAction.REDUCE_INTENSITY,
            )
            drain == DrainLevel.HIGH && readiness == ReadinessLevel.AVERAGE && isLateSession -> CoachMessageComponents(
                severity = CoachSeverity.WARNING,
                title = "Cerca del límite — serie final de esta fase",
                body = "Estás drenado pero tu cuerpo responde. Si la próxima serie se siente pesada, no insistas. Termina fuerte pero seguro.",
                action = CoachAction.REDUCE_INTENSITY,
            )
            drain == DrainLevel.HIGH && readiness == ReadinessLevel.HIGH && isLateSession -> CoachMessageComponents(
                severity = CoachSeverity.INFO,
                title = "Drenado pero listo",
                body = "Tu batería está baja pero tu readiness es alta. Recuperas bien. La última serie cuenta.",
                action = CoachAction.STAY_THE_COURSE,
            )
            drain == DrainLevel.HIGH && readiness == ReadinessLevel.HIGH && !isLateSession -> CoachMessageComponents(
                severity = CoachSeverity.INFO,
                title = "Buen ritmo de trabajo",
                body = "Estás drenando pero tu recuperación es excelente. Sigue con el plan, tu cuerpo responde bien.",
                action = CoachAction.STAY_THE_COURSE,
            )
            drain == DrainLevel.AVERAGE && readiness == ReadinessLevel.HIGH && isLateSession -> CoachMessageComponents(
                severity = CoachSeverity.SUCCESS,
                title = "Excelente sesión",
                body = "Drenado moderado con readiness alta. Estás navegando la sesión con eficiencia. ¡A la carga!",
                action = CoachAction.STAY_THE_COURSE,
            )
            drain == DrainLevel.AVERAGE && readiness == ReadinessLevel.LOW -> CoachMessageComponents(
                severity = CoachSeverity.WARNING,
                title = "Recuperación limitada detectada",
                body = "Tu cuerpo no está al 100% hoy. Ajusta la intensidad si algo no se siente bien.",
                action = CoachAction.REDUCE_INTENSITY,
            )
            drain == DrainLevel.LOW && readiness == ReadinessLevel.LOW -> CoachMessageComponents(
                severity = CoachSeverity.INFO,
                title = "Sesión ligera — oportunidad",
                body = "Baja fatiga y readiness moderada. Ideal para trabajar técnica o accesibilidad. ¡Aprovecha!",
                action = CoachAction.STAY_THE_COURSE,
            )
            drain == DrainLevel.LOW && readiness == ReadinessLevel.HIGH -> CoachMessageComponents(
                severity = CoachSeverity.SUCCESS,
                title = "Día óptimo para max effort",
                body = "Baja fatiga y máxima disponibilidad. Ideal para series pesadas o probar RMs.",
                action = CoachAction.STAY_THE_COURSE,
            )
            drain == DrainLevel.XS && readiness == ReadinessLevel.HIGH -> CoachMessageComponents(
                severity = CoachSeverity.SUCCESS,
                title = "¡Energía al máximo!",
                body = "Sesión nueva, cuerpo fresco. Perfecto para series pesadas o explorar nuevos límites.",
                action = CoachAction.STAY_THE_COURSE,
            )
            drain == DrainLevel.XS && readiness == ReadinessLevel.XS -> CoachMessageComponents(
                severity = CoachSeverity.WARNING,
                title = "Disponibilidad reducida",
                body = "Readiness muy baja. Si es posible, considera reprogramar sesiones pesadas para otro día.",
                action = CoachAction.SKIP_EXERCISE,
            )
            drain == DrainLevel.HIGH && readiness == ReadinessLevel.AVERAGE && !isLateSession -> CoachMessageComponents(
                severity = CoachSeverity.INFO,
                title = "Mitad de sesión — evalúa cómo vas",
                body = "Has drenado bastante. Si aún te quedan series pesadas, considera reducir un 5-10% en las últimas.",
                action = CoachAction.EXTEND_REST,
            )
            else -> CoachMessageComponents(
                severity = CoachSeverity.INFO,
                title = "Seguimiento en tiempo real",
                body = "El motor AUGE monitoriza tu fatiga continuamente. Los ajustes de descanso se aplican automáticamente.",
                action = null,
            )
        }
    }

    private data class CoachMessageComponents(
        val severity: CoachSeverity,
        val title: String,
        val body: String,
        val action: CoachAction?,
    )

    fun getMessage(
        weightedDrainPct: Double,
        readinessScore: Int?,
        sessionProgress: Double,
    ): CoachMessage {
        val drainLevel = when {
            weightedDrainPct >= 8.0 -> DrainLevel.HIGH
            weightedDrainPct >= 4.0 -> DrainLevel.AVERAGE
            weightedDrainPct >= 1.5 -> DrainLevel.LOW
            else -> DrainLevel.XS
        }
        val readinessLevel = when (readinessScore) {
            null, in 80..100 -> ReadinessLevel.HIGH
            in 60..79 -> ReadinessLevel.AVERAGE
            in 40..59 -> ReadinessLevel.LOW
            else -> ReadinessLevel.XS
        }
        val phase = when {
            sessionProgress < 0.33 -> SessionPhase.EARLY
            sessionProgress < 0.66 -> SessionPhase.MID
            else -> SessionPhase.LATE
        }
        return COACH_MESSAGES[CoachKey(drainLevel, readinessLevel, phase)]
            ?: COACH_MESSAGES.getValue(CoachKey(DrainLevel.LOW, ReadinessLevel.AVERAGE, SessionPhase.MID))
    }

    fun getReadinessScore(
        neural: Int?,
        spinal: Int?,
        muscular: Int?,
    ): Int? {
        val values = listOfNotNull(neural, spinal, muscular)
        return if (values.isNotEmpty()) values.average().toInt() else null
    }
}

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Workout/WorkoutCoachMessages.swift
import Foundation

struct CoachMessage {
    let key: String
    let title: String
    let body: String
    let severity: CoachSeverity
    let action: CoachAction?
}

enum CoachSeverity { case info, warning, danger, success }
enum CoachAction { case reduceIntensity, skipExercise, extendRest, stayTheCourse }

struct WorkoutCoachMessages {
    enum DrainLevel { case xs, low, average, high }
    enum ReadinessLevel { case xs, low, average, high }
    private enum SessionPhase { case early, mid, late }

    private struct CoachKey: Hashable {
        let drain: DrainLevel
        let readiness: ReadinessLevel
        let sessionProgress: SessionPhase
    }

    private struct CoachMessageComponents {
        let severity: CoachSeverity
        let title: String
        let body: String
        let action: CoachAction?
    }

    private static let coachMessages: [CoachKey: CoachMessage] = {
        var map = [CoachKey: CoachMessage]()
        for drain in [DrainLevel.xs, .low, .average, .high] {
            for readiness in [ReadinessLevel.xs, .low, .average, .high] {
                for phase in [SessionPhase.early, .mid, .late] {
                    let resolved = resolve(drain: drain, readiness: readiness, phase: phase)
                    map[CoachKey(drain: drain, readiness: readiness, sessionProgress: phase)] = CoachMessage(
                        key: "\(drain)_\(readiness)_\(phase)",
                        title: resolved.title,
                        body: resolved.body,
                        severity: resolved.severity,
                        action: resolved.action
                    )
                }
            }
        }
        return map
    }()

    private static func resolve(drain: DrainLevel, readiness: ReadinessLevel, phase: SessionPhase) -> CoachMessageComponents {
        let isHighDrain = drain == .high || drain == .average
        let isLowReadiness = readiness == .xs || readiness == .low
        let isLateSession = phase == .late

        switch (isHighDrain, isLowReadiness, isLateSession, drain, readiness, phase) {
        case (true, true, true, _, _, _):
            return CoachMessageComponents(
                severity: .danger,
                title: "Cuidado: fatiga acumulada + baja disponibilidad",
                body: "Has acumulado fatiga significativa y tu readiness está baja. Considera terminar ejercicios principales aquí y guardar energía.",
                action: .reduceIntensity
            )
        case (true, true, false, _, _, _):
            return CoachMessageComponents(
                severity: .warning,
                title: "Fatiga en aumento, readiness limitada",
                body: "El sistema nervioso está trabajando fuerte y la recuperación de hoy no es óptima. Prioriza las series más importantes.",
                action: .reduceIntensity
            )
        case (_, _, _, .high, .average, .late):
            return CoachMessageComponents(
                severity: .warning,
                title: "Cerca del límite — serie final de esta fase",
                body: "Estás drenado pero tu cuerpo responde. Si la próxima serie se siente pesada, no insistas. Termina fuerte pero seguro.",
                action: .reduceIntensity
            )
        case (_, _, _, .high, .high, .late):
            return CoachMessageComponents(
                severity: .info,
                title: "Drenado pero listo",
                body: "Tu batería está baja pero tu readiness es alta. Recuperas bien. La última serie cuenta.",
                action: .stayTheCourse
            )
        case (_, _, _, .high, .high, _):
            return CoachMessageComponents(
                severity: .info,
                title: "Buen ritmo de trabajo",
                body: "Estás drenando pero tu recuperación es excelente. Sigue con el plan, tu cuerpo responde bien.",
                action: .stayTheCourse
            )
        case (_, _, _, .average, .high, .late):
            return CoachMessageComponents(
                severity: .success,
                title: "Excelente sesión",
                body: "Drenado moderado con readiness alta. Estás navegando la sesión con eficiencia. ¡A la carga!",
                action: .stayTheCourse
            )
        case (_, _, _, .average, .low, _):
            return CoachMessageComponents(
                severity: .warning,
                title: "Recuperación limitada detectada",
                body: "Tu cuerpo no está al 100% hoy. Ajusta la intensidad si algo no se siente bien.",
                action: .reduceIntensity
            )
        case (false, true, _, .low, _, _):
            return CoachMessageComponents(
                severity: .info,
                title: "Sesión ligera — oportunidad",
                body: "Baja fatiga y readiness moderada. Ideal para trabajar técnica o accesibilidad. ¡Aprovecha!",
                action: .stayTheCourse
            )
        case (false, false, _, .low, .high, _):
            return CoachMessageComponents(
                severity: .success,
                title: "Día óptimo para max effort",
                body: "Baja fatiga y máxima disponibilidad. Ideal para series pesadas o probar RMs.",
                action: .stayTheCourse
            )
        case (false, false, _, .xs, .high, _):
            return CoachMessageComponents(
                severity: .success,
                title: "¡Energía al máximo!",
                body: "Sesión nueva, cuerpo fresco. Perfecto para series pesadas o explorar nuevos límites.",
                action: .stayTheCourse
            )
        case (false, true, _, .xs, .xs, _):
            return CoachMessageComponents(
                severity: .warning,
                title: "Disponibilidad reducida",
                body: "Readiness muy baja. Si es posible, considera reprogramar sesiones pesadas para otro día.",
                action: .skipExercise
            )
        case (true, false, _, .high, .average, _):
            return CoachMessageComponents(
                severity: .info,
                title: "Mitad de sesión — evalúa cómo vas",
                body: "Has drenado bastante. Si aún te quedan series pesadas, considera reducir un 5-10% en las últimas.",
                action: .extendRest
            )
        default:
            return CoachMessageComponents(
                severity: .info,
                title: "Seguimiento en tiempo real",
                body: "La sesión ajusta el descanso según tu fatiga registrada.",
                action: nil
            )
        }
    }

    static func getMessage(
        weightedDrainPct: Double,
        readinessScore: Int?,
        sessionProgress: Double
    ) -> CoachMessage {
        let drainLevel: DrainLevel = {
            switch weightedDrainPct {
            case 8.0...: return .high
            case 4.0..<8.0: return .average
            case 1.5..<4.0: return .low
            default: return .xs
            }
        }()
        let readinessLevel: ReadinessLevel = {
            guard let score = readinessScore else { return .high }
            switch score {
            case 80...100: return .high
            case 60..<80: return .average
            case 40..<60: return .low
            default: return .xs
            }
        }()
        let phase: SessionPhase = {
            switch sessionProgress {
            case ..<0.33: return .early
            case 0.33..<0.66: return .mid
            default: return .late
            }
        }()
        return coachMessages[CoachKey(drain: drainLevel, readiness: readinessLevel, sessionProgress: phase)]
            ?? coachMessages[CoachKey(drain: .low, readiness: .average, sessionProgress: .mid)]!
    }

    static func getReadinessScore(neural: Int?, spinal: Int?, muscular: Int?) -> Int? {
        let values = [neural, spinal, muscular].compactMap { $0 }
        guard !values.isEmpty else { return nil }
        return values.reduce(0, +) / values.count
    }
}
